test/ts/Component.tsx

```diff
@@ -1,29 +1,70 @@
-import { h, Component } from "hyperapp"
+import {
+  h,
+  app,
+  LazyComponent,
+  Component,
+  View,
+  ActionType,
+  ActionsType
+} from "hyperapp"
 
 interface State {
   count: number
 }
 
+const state: State = { count: 10 }
+
 interface Actions {
-  down(): State
-  up(value: number): State
+  down: ActionType<State, Actions>
+  up: ActionType<State, Actions>
+}
+
+const actions: ActionsType<State, Actions> = {
+  down: () => state => ({ count: state.count - 1 }),
+  up: (value: number) => state => ({
+    count: state.count + value
+  })
 }
 
-const Counter: Component<{ state: State; actions: Actions }> = ({
-  state,
-  actions
-}) => (
+interface ComponentParams {
+  count: number
+  up: () => void
```

Comment:
Actually, an `action` returns the new `State`. 

Why is this not typed as `Action`?

Reply:
Simply as its a property function passed to a component (ie callback if you like). it takes no parameters and returns nothing. The parent conmpenent has the logic in . This is a common pattern in React etc. You have a parent "component" with logic and state and dump child components. 

Perhaps we should not be introducing patterns here.

Also it might not be a hyperapp pattern you could pass a global action directly i guess , I just baulked at passing all Actions as you can;t tel which are used by the component form it's signature

Reply:
> it takes no parameters and returns nothing
Where you use the component it clearly passes a WiredAction. Especially the 'down' action, but also the 'up' action returns the new state. 

```
<Counter count={state.count} up={() => actions.up(2)} down={actions.down} />
```

> Perhaps we should not be introducing patterns here.
What pattern do you mean? I'm simply referring to the typings, they're incorrect this way.

Reply:
That is an error. Good spot. I intended 

```
<Counter count={state.count} up={() => {actions.up(2)}} down={actions.down} />
```

I use {....} to indicate no return - ie invoked for side effects only

This [pattern](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0)

Reply:
> Where you use the component it clearly passes a WiredAction. Especially the 'down' action, but also the 'up' action returns the new state.

The callee has no idea it is an action - It just gets a function. That's intentional decoupling. passing directly is a short cut for

```
down={()=>{actions.down()}}
```

The call does not pass any args and ignores the return so type script is OK with it. You might want to be explicit about it an use long for